readme.txt
=== WooCommerce Product Bundles ===

Contributors: SomewhereWarm
Tags: woocommerce, product, bundle, bundles, kits, simple, variable, configurable
Requires at least: 4.4
Tested up to: 5.8
Stable tag: 6.12.0
WC requires at least: 3.6
WC tested up to: 5.7
License: GNU General Public License v3.0
License URI: http://www.gnu.org/licenses/gpl-3.0.html

Offer product bundles, bulk discount packages, and assembled products.

== Description ==

Create assembled products and product bundles by grouping simple, variable and subscription products.

Looking for help? Read the full documentation [here](http://docs.woocommerce.com/document/bundles/).


changelog.txt
*** WooCommerce Product Bundles ***

2021.09.01 - version 6.12.0
* Important - Product Bundles now requires WooCommerce 3.6+.
* Feature - Added integration with the Revenue Analytics of Composite Products.
* Feature - Made it possible to remove optional bundled items in the Cart page.
* Fix - Make sure original bundled item price is displayed when the bundled item has a discount.
* Fix - Prevent bundled products from being reported as 'on backorder' in bundles, unless backordering requires notification.
* Fix - Allow bundle-sells discounts granted by Composite Products, bundled products, and composited products to be applied to other products in the cart.
* Fix - Update 'Edit in Cart' query string to include bundle quantity.
* Fix - Fixed an infinite loop affecting discounts when using Memberships. Thanks @Alex for pointing this out!
* Fix - Fixed a fatal error that is triggered by invalid 'wc_pb_notice_options' values.
* Fix - Fixed PHP notice triggered when a non-numeric value is used in 'WC_Product_Bundle::calculate_price'.
* Tweak - Bundled items are now displayed under composited Bundles in the mini cart when the 'Item Grouping > Flat' is selected.

2021.07.12 - version 6.11.1
* Feature - Added 'bundle_stock_status' and 'bundle_stock_quantity' REST API fields (read only).
* Fix - Fixed an indexing issue affecting the 'wc_order_bundle_lookup' table.

2021.07.08 - version 6.11.0
* Feature - Re-introduce 'Virtual' checkbox for Product Bundles. When saving a Bundle as Virtual, all physical bundled products will be treated as Virtual, too.
* Tweak - Add 'woocommerce_cart_item_removed' action to 'WC_PB_Cart'.
* Tweak - Made bundled product IDs clickable under 'Product Data > Bundled Products'.
* Tweak - Added 'add-to-cart' parameter to submit button.

2021.06.24 - version 6.10.3
* Fix - Prevent 'get_transient' calls from firing when bundled item map cache is not in use.
* Fix - Max item quantity not taken into account when calculating the stock status and remaining stock for bundles with 'Min Bundle Size' < 0.
* Tweak - Dropped support for exporting/downloading revenue reports under 'Analytics > Bundles'. Requires additional extensibility points in WooCommerce core to be supported reliably.

2021.06.15 - version 6.10.2
* Fix - Respect 'woocommerce_admin_disabled' filter when using WooCommerce Admin 2.3+

2021.06.14 - version 6.10.1
* Fix - Fixed an issue that could cause remaining bundle stock to be calculated incorrectly when bundling variable products that manage stock at parent level.
* Tweak - Display selected quantity in bundled item titles when item quantity is configurable.

2021.06.01 - version 6.10.0
* Feature - Send stock notifications for bundles that run low in stock or out of stock due to bundled product stock changes.
* Feature - View remaining bundles in product bundle pages, even when 'Product Data > Inventory > Manage Stock' is disabled at bundle level. Applicable to bundles that contain at least one bundled item that manages stock.
* Feature - View remaining bundles in 'Products' list table. Calculated based on minimum bundled item quantities. Applicable to bundles that contain at least one bundled item that manages stock.
* Fix - Prevent bundle cart/order item prices/subtotals from being hidden when bundled item prices/subtotals are not aggregate and the parent item price/subtotal is zero.
* Fix - Always display bundled items in orders where the parent bundle doesn't exist - for example, in 'My Account > Subscriptions'.

2021.05.26 - version 6.9.0
* Feature - Introduced new 'Revenue' and 'Stock' reports under 'Analytics > Bundles'. The new 'Stock' report replaces the one previously found under 'WooCommerce > Reports > Stock > Insufficient Stock'.
* Feature - Added support for variations with 'Name Your Price' inputs.
* Fix - Fixed an issue that prevented bundled item data from importing correctly when overriding bundled item titles/descriptions and using double quotes.
* Fix - Prevent 'PHP Recoverable fatal error' during checkout in some edge cases.
* Fix - Fixed the bundled items checkboxes layout in the Twenty-Twenty-One theme.
* Fix - Ensure that Membership discounts apply on bundled items if the user adds an eligible product to the cart.
* Tweak - Improve stock sync logging to record 'full queue' events.

2021.03.22 - version 6.8.0
* Tweak - Throttle dispatches of stock sync tasks to lower resource utilization on high traffic sites.
* Fix - Make sure visibility taxonomy is properly updated when using the 'WC_PB_DEBUG_STOCK_SYNC' and 'WC_PB_DEBUG_STOCK_PARENT_SYNC' constants.

2021.03.06 - version 6.7.7
* Fix - Fixed a fatal error caused by bundling subscription-type products while 'Priced Individually' is disabled.
* Fix - Fixed a tax-related infinite loop caused by a regression in version 6.7.6.

2021.03.04 - version 6.7.6
* Dev - Added 'woocommerce_after_bundle_price', 'woocommerce_before_bundle_availability', and 'woocommerce_before_bundle_add_to_cart_button' hooks.
* Tweak - Increased the discounted price precision relative to the store precision to prevent errors such as: $10 x 10 = $100 before discount, and ( $9.5 rounded to $10 ) x 10 = $100 after 5% discount.
* Fix - Bundled item totals are now rounded before calculating the Bundle total, to prevent inconsistencies between the total calculated via JS and the catalog price of a Bundle.
* Fix - Fixed 'Too few arguments to function WC_PB_Order::get_product_from_item' error when reading order data via the REST API.

2021.02.22 - version 6.7.5
* Fix - Prevented a fatal error thrown while initializing the plugin if WooCommerce is not activated.
* Fix - Prevented a REST API fatal error in POST/PUT context in some scenarios.

2021.02.15 - version 6.7.4
* Tweak - Use the core 'WC_Background_Process' library for synchronizing stock data between bundled products and bundles in the background.
* Tweak - Removed quantity multipliers from bundled product titles when using the Tabular layout.

woocommerce-product-bundles.php
<?php
/**
* Plugin Name: WooCommerce Product Bundles
* Plugin URI: https://woocommerce.com/products/product-bundles/
* Description: Offer product bundles, bulk discount packages, and assembled products.
* Version: 6.12.0
* Author: SomewhereWarm
* Author URI: https://somewherewarm.com/
*
* Woo: 18716:fbca839929aaddc78797a5b511c14da9
*
* Text Domain: woocommerce-product-bundles
* Domain Path: /languages/
*
* Requires PHP: 5.6
*
* Requires at least: 4.4
* Tested up to: 5.8
*
* WC requires at least: 3.6
* WC tested up to: 5.7
*
* Copyright: © 2017-2021 SomewhereWarm SMPC.
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/**
 * Main plugin class.
 *
 * @class    WC_Bundles
 * @version  6.12.0
 */
class WC_Bundles {

	public $version  = '6.12.0';
	public $required = '3.6.0';

	/**
	 * The single instance of the class.
	 * @var WC_Bundles
	 *
	 * @since 4.11.4
	 */
	protected static $_instance = null;

	/**
	 * Main WC_Bundles instance. Ensures only one instance of WC_Bundles is loaded or can be loaded - @see 'WC_PB()'.
	 *
	 * @static
	 * @return WC_Bundles
	 * @since  4.11.4
	 */
	public static function instance() {
		if ( is_null( self::$_instance ) ) {
			self::$_instance = new self();
		}
		return self::$_instance;
	}

	/**
	 * Cloning is forbidden.
	 *
	 * @since 4.11.4
	 */
	public function __clone() {
		_doing_it_wrong( __FUNCTION__, __( 'Foul!', 'woocommerce-product-bundles' ), '4.11.4' );
	}

	/**
	 * Unserializing instances of this class is forbidden.
	 *
	 * @since 4.11.4
	 */
	public function __wakeup() {
		_doing_it_wrong( __FUNCTION__, __( 'Foul!', 'woocommerce-product-bundles' ), '4.11.4' );
	}

	/**
	 * Make stuff.
	 */
	protected function __construct() {
		// Entry point.
		add_action( 'plugins_loaded', array( $this, 'initialize_plugin' ), 9 );
	}

	/**
	 * Auto-load in-accessible properties.
	 *
	 * @param  mixed  $key
	 * @return mixed
	 */
	public function __get( $key ) {
		if ( in_array( $key, array( 'compatibility', 'modules', 'cart', 'order', 'display' ) ) ) {
			$classname = 'WC_PB_' . ucfirst( $key );
			return call_user_func( array( $classname, 'instance' ) );
		}
	}

	/**
	 * Plugin URL getter.
	 *
	 * @return string
	 */
	public function plugin_url() {
		return untrailingslashit( plugins_url( '/', __FILE__ ) );
	}

	/**
	 * Plugin path getter.
	 *
	 * @return string
	 */
	public function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	/**
	 * Plugin base path name getter.
	 *
	 * @return string
	 */
	public function plugin_basename() {
		return plugin_basename( __FILE__ );
	}

	/**
	 * Indicates whether the plugin has been fully initialized.
	 *
	 * @since  6.2.0
	 *
	 * @return boolean
	 */
	public function plugin_initialized() {
		return class_exists( 'WC_PB_Helpers' );
	}

	/**
	 * Define constants if not present.
	 *
	 * @since  6.2.0
	 *
	 * @return boolean
	 */
	protected function maybe_define_constant( $name, $value ) {
		if ( ! defined( $name ) ) {
			define( $name, $value );
		}
	}

	/**
	 * Plugin version getter.
	 *
	 * @since  5.8.0
	 *
	 * @param  boolean  $base
	 * @param  string   $version
	 * @return string
	 */
	public function plugin_version( $base = false, $version = '' ) {

		$version = $version ? $version : $this->version;

		if ( $base ) {
			$version_parts = explode( '-', $version );
			$version       = sizeof( $version_parts ) > 1 ? $version_parts[ 0 ] : $version;
		}

		return $version;
	}

	/**
	 * Fire in the hole!
	 */
	public function initialize_plugin() {

		$this->define_constants();
		$this->maybe_create_store();

		// WC version sanity check.
		if ( ! function_exists( 'WC' ) || version_compare( WC()->version, $this->required ) < 0 ) {
			/* translators: Version */
			$notice = sprintf( __( 'WooCommerce Product Bundles requires at least WooCommerce <strong>%s</strong>.', 'woocommerce-product-bundles' ), $this->required );
			require_once( WC_PB_ABSPATH . 'includes/admin/class-wc-pb-admin-notices.php' );
			WC_PB_Admin_Notices::add_notice( $notice, 'error' );
			return false;
		}

		// PHP version check.
		if ( ! function_exists( 'phpversion' ) || version_compare( phpversion(), '5.6.20', '<' ) ) {
			/* translators: %1$s: Version %, %2$s: Update PHP doc URL */
			$notice = sprintf( __( 'WooCommerce Product Bundles requires at least PHP <strong>%1$s</strong>. Learn <a href="%2$s">how to update PHP</a>.', 'woocommerce-product-bundles' ), '5.6.20', $this->get_resource_url( 'update-php' ) );
			require_once( WC_PB_ABSPATH . 'includes/admin/class-wc-pb-admin-notices.php' );
			WC_PB_Admin_Notices::add_notice( $notice, 'error' );
			return false;
		}

		$this->includes();

		WC_PB_Compatibility::instance();
		WC_PB_Modules::instance();

		WC_PB_Cart::instance();
		$this->modules->load_components( 'cart' );

		WC_PB_Order::instance();
		$this->modules->load_components( 'order' );

		WC_PB_Display::instance();
		$this->modules->load_components( 'display' );

		// Load translations hook.
		add_action( 'init', array( $this, 'load_translation' ) );
	}

	/**
	 * Constants.
	 */
	public function define_constants() {

		$this->maybe_define_constant( 'WC_PB_VERSION', $this->version );
		$this->maybe_define_constant( 'WC_PB_SUPPORT_URL', 'https://woocommerce.com/my-account/marketplace-ticket-form/' );
		$this->maybe_define_constant( 'WC_PB_ABSPATH', trailingslashit( plugin_dir_path( __FILE__ ) ) );

		/*
		 * Available debug constants:
		 *
		 * 'WC_PB_DEBUG_STOCK_CACHE' - Used to disable bundled item stock caching.
		 *
		 * 'WC_PB_DEBUG_STOCK_SYNC' - Used to disable bundled item stock syncing in the background.
		 *
		 * 'WC_PB_DEBUG_STOCK_PARENT_SYNC' - Used to disable stock status and visibility syncing for bundle containers.
		 *
		 * 'WC_PB_DEBUG_TRANSIENTS' - Used to disable transients caching.
		 *
		 * 'WC_PB_DEBUG_OBJECT_CACHE' - Used to disable object caching.
		 *
		 * 'WC_PB_DEBUG_RUNTIME_CACHE' - Used to disable runtime object caching.
		 */

		if ( defined( 'WC_PB_DEBUG_STOCK_CACHE' ) ) {
			/**
			 * 'WC_PB_DEBUG_STOCK_SYNC' constant.
			 *
			 * Used to disable bundled product stock meta syncing for bundled items.
			 */
			$this->maybe_define_constant( 'WC_PB_DEBUG_STOCK_SYNC', true );
		}

		if ( defined( 'WC_PB_DEBUG_STOCK_SYNC' ) || ! function_exists( 'WC' ) || version_compare( WC()->version, '3.3.0' ) < 0 ) {
			/**
			 * 'WC_PB_DEBUG_STOCK_PARENT_SYNC' constant.
			 *
			 * Used to disable stock status and visibility syncing for bundles.
			 * Requires the 'WC_Background_Process' class introduced in WC 3.3.
			 */
			$this->maybe_define_constant( 'WC_PB_DEBUG_STOCK_PARENT_SYNC', true );
		}
	}

	/**
	 * A simple dumb datastore for sharing information accross our plugins.
	 *
	 * @since  6.3.0
	 *
	 * @return void
	 */
	private function maybe_create_store() {
		if ( ! isset( $GLOBALS[ 'sw_store' ] ) ) {
			$GLOBALS[ 'sw_store' ] = array();
		}
	}

	/**
	 * Includes.
	 */
	public function includes() {

		// Extensions compatibility functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/compatibility/class-wc-pb-compatibility.php' );

		// Modules.
		require_once( WC_PB_ABSPATH . 'includes/modules/class-wc-pb-modules.php' );

		// Data classes.
		require_once( WC_PB_ABSPATH . 'includes/data/class-wc-pb-data.php' );

		// Install.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-install.php' );

		// Functions (incl deprecated).
		require_once( WC_PB_ABSPATH . 'includes/wc-pb-functions.php' );
		require_once( WC_PB_ABSPATH . 'includes/wc-pb-deprecated-functions.php' );

		// Helper functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-helpers.php' );

		// Data syncing between products and bundled items.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-db-sync.php' );

		// Product price filters and price-related functions.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-product-prices.php' );

		// Bundled Item class.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-bundled-item.php' );

		// Product Bundle class.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-product-bundle.php' );

		// Stock mgr class.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-stock-manager.php' );

		// Cart-related functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-cart.php' );

		// Order-related functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-order.php' );

		// Order-again functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-order-again.php' );

		// Coupon-related functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-coupon.php' );

		// Front-end filters and templates.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-display.php' );

		// Front-end AJAX handlers.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-ajax.php' );

		// REST API hooks.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-rest-api.php' );

		// Notices handling.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-notices.php' );

		// Stock notifications handling.
		require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-stock-notifications.php' );

		// Admin includes.
		if ( is_admin() ) {
			$this->admin_includes();
		}

		// Analytics.
		require_once( WC_PB_ABSPATH . 'includes/admin/analytics/class-wc-pb-admin-analytics.php' );

		// WP-CLI includes.
		if ( defined( 'WP_CLI' ) && WP_CLI ) {
			require_once( WC_PB_ABSPATH . 'includes/class-wc-pb-cli.php' );
		}
	}

	/**
	 * Admin & AJAX functions and hooks.
	 */
	public function admin_includes() {

		// Admin notices handling.
		require_once( WC_PB_ABSPATH . 'includes/admin/class-wc-pb-admin-notices.php' );

		// Admin functions and hooks.
		require_once( WC_PB_ABSPATH . 'includes/admin/class-wc-pb-admin.php' );
	}

	/**
	 * Load textdomain.
	 */
	public function load_translation() {
		load_plugin_textdomain( 'woocommerce-product-bundles', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
	}

	/**
	 * Returns URL to a doc or support resource.
	 *
	 * @since  6.3.0
	 *
	 * @param  string  $handle
	 * @return string
	 */
	public function get_resource_url( $handle ) {

		$resource = false;

		if ( 'pricing-options' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/bundles/bundles-configuration/#pricing';
		} elseif ( 'shipping-options' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/bundles/bundles-configuration/#shipping';
		} elseif ( 'update-php' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/how-to-update-your-php-version/';
		} elseif ( 'docs-contents' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/bundles/';
		} elseif ( 'max-input-vars' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/bundles/bundles-faq/#faq_bundled_items_dont_save';
		} elseif ( 'updating' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/how-to-update-woocommerce/';
		} elseif ( 'min-max' === $handle ) {
			$resource = 'https://wordpress.org/plugins/product-bundles-minmax-items-for-woocommerce/';
		} elseif ( 'bulk-discounts' === $handle ) {
			$resource = 'https://wordpress.org/plugins/product-bundles-bulk-discounts-for-woocommerce/';
		} elseif ( 'analytics-revenue' === $handle ) {
			$resource = 'https://docs.woocommerce.com/document/bundles/bundles-faq/#faq_analytics_bundles_revenue';
		} elseif ( 'ticket-form' === $handle ) {
			$resource = WC_PB_SUPPORT_URL;
		}

		return $resource;
	}

	/*
	|--------------------------------------------------------------------------
	| Deprecated methods.
	|--------------------------------------------------------------------------
	*/

	public function woo_bundles_plugin_url() {
		_deprecated_function( __METHOD__ . '()', '5.0.0', __CLASS__ . '::plugin_url()' );
		return $this->plugin_url();
	}
	public function woo_bundles_plugin_path() {
		_deprecated_function( __METHOD__ . '()', '5.0.0', __CLASS__ . '::plugin_path()' );
		return $this->plugin_path();
	}
}

/**
 * Returns the main instance of WC_Bundles to prevent the need to use globals.
 *
 * @since  4.11.4
 * @return WC_Bundles
 */
function WC_PB() {
	return WC_Bundles::instance();
}

$GLOBALS[ 'woocommerce_bundles' ] = WC_PB();
